### DISCUSSION TITLE
[Question] Integration with Rustaceanvim

### DISCUSSION THREAD

**Discussion by pi314ever** [pr2502/ra-multiplex#79] [Q&A]
How do I integrate this with [rustaceanvim](https://github.com/mrcjkb/rustaceanvim)? It has many more features that I would love to have that builds on top of rust analyzer, but I cannot figure out an interface between ra-multiplex and rustaceanvim. 

I have tried `ln -s ra-multiplex rust-analyzer` to see if it can drop-in replace at least the client side, but it doesn't work due to differing APIs. Specifically, it complained that it could not use `--log-file` flag. 

Any help on this would be greatly appreciated!

**Comment by EliteTK** [comment]
Having had a look at the rustaceanvim help (`:help rustaceanvim.config` I believe) there's a rustaceanvim configuration object which contains a settings object which can override cmd. It doesn't look like this can take the result of `vim.lsp.rpc.connect()` but you can try that first (as per the neovim example). If it doesn't work, set it to the path of "ra-multiplex".

While undocumented, rustaceanvim appears to also look for an init_options in the same place (as per https://github.com/mrcjkb/rustaceanvim/blob/29f42cc149f915d771c550b6dfe7c788d856cf04/lua/rustaceanvim/lsp/init.lua#L112 ).

Once you set those two things, it hopefully should work.

**Comment by pr2502** [comment]
> I have tried ln -s ra-multiplex rust-analyzer to see if it can drop-in replace at least the client side, but it doesn't work due to differing APIs. Specifically, it complained that it could not use --log-file flag.

Another option for this approach would be to write a shell script instead of symlinking. Ignore the passed options and add `--server-path` option if you need to set it.

Something like `/usr/local/bin/rust-analyzer`:
```bash
#!/bin/bash
exec /path/to/ra-multiplex client --server-path /path/to/real/rust-analyzer
```

Then make sure that this shell script appears in `PATH` before the real rust-analyzer for your editor. You can use `which rust-analyzer` command in the same shell you'd spawn your editor from to see what it finds first.

**Comment by alepez** [comment]
With this configuration it works:

```
vim.g.rustaceanvim = {
  server = {
    cmd = function()
      return { "ra-multiplex" }
    end,
```

**Comment by mercxry** [comment]
Since [rustaceanvim v5.20](https://github.com/mrcjkb/rustaceanvim/releases/tag/v5.20.0) there is auto-discovery for `ra-multiplex` if it's running in localhost with the default port and the `server.cmd` option is **not** set, so now you don't have to specify it anymore in the config and you have an automatic fallback to `rust-analyzer` if `ra-multiplex` is not running.
